apps/passport-client/src/validateState.ts

```diff
@@ -0,0 +1,241 @@
+} from "@pcd/semaphore-identity-pcd";
+import { Identity } from "@semaphore-protocol/identity";
+import { appConfig } from "./appConfig";
+import { loadSelf } from "./localstorage";
+import { AppState } from "./state";
+
+/**
+ * Returns `true` if {@link validateInitialAppState} returns no errors, and `false`
+ * otherwise. In the case that errors were encountered, uploads an error report to
+ * the Zupass backend.
+ */
+export function validateAndLogInitialAppState(
+  tag: string,
+  state: AppState
+): boolean {
+  const validationErrors = validateInitialAppState(tag, state);
+
+  if (validationErrors.errors.length > 0) {
+    logValidationErrors(validationErrors);
+    return false;
+  }
+
+  return true;
+}
+
+/**
+ * Returns `true` if {@link validateRunningAppState} returns no errors, and `false`
+ * otherwise. In the case that errors were encountered, uploads an error report to
+ * the Zupass backend.
+ */
+export function validateAndLogRunningAppState(
+  tag: string,
+  self: User | undefined,
+  identity: Identity | undefined,
+  pcds: PCDCollection | undefined,
+  forceCheckPCDs?: boolean
+): boolean {
+  const validationErrors = validateRunningAppState(
+    tag,
+    self,
+    identity,
+    pcds,
+    forceCheckPCDs
+  );
+
+  if (validationErrors.errors.length > 0) {
+    logValidationErrors(validationErrors);
+    return false;
+  }
+
+  return true;
+}
+
+/**
+ * Validates state of a Zupass application that is just starting up (using the
+ * {@link validateRunningAppState} alongside some extra ones), and returns an {@link ErrorReport}.
+ */
+export function validateInitialAppState(
+  tag: string,
+  appState: AppState | undefined
+): ErrorReport {
+  return {
+    errors: getInitialAppStateValidationErrors(appState),
+    userUUID: appState?.self?.uuid,
+    tag
+  };
+}
+
+/**
+ * Validates state of a running Zupass application and returns an {@link ErrorReport}.
+ */
+export function validateRunningAppState(
+  tag: string,
+  self: User | undefined,
+  identity: Identity | undefined,
+  pcds: PCDCollection | undefined,
+  forceCheckPCDs?: boolean
+): ErrorReport {
+  return {
+    errors: getRunningAppStateValidationErrors(
+      self,
+      identity,
+      pcds,
+      forceCheckPCDs
+    ),
+    userUUID: self?.uuid,
+    tag
+  };
+}
+
+/**
+ * Validates state of a Zupass application that is just starting up. Uses
+ * {@link getRunningAppStateValidationErrors}, and performs some extra checks
+ * on top of it.
+ */
+export function getInitialAppStateValidationErrors(state: AppState): string[] {
+  const errors = [
+    ...getRunningAppStateValidationErrors(
+      state.self,
+      state.identity,
+      state.pcds
+    )
+  ];
+
+  // this case covers a logged in user. the only way the app can get a 'self'
+  // is by requesting one from the server, to do which one has to be logged in.
+  if (state.self) {
+    // TODO: any extra checks that need to happen on immediate app startup should
```

Comment:
When I discovered the case where the encryption key is optional, did that really make this list empty?  Sorry for all the back-and-forth over it if so.  There may be other AppState invariants not related to sync to put here in future, but I'm not thinking of any right now.

Reply:
No worries!